--- REL_17_4/src/backend/commands/async.c
+++ pglite-REL_17_4/src/backend/commands/async.c
@@ -1651,8 +1651,12 @@
 		 * NotifyQueueLock; which is unlikely but certainly possible. So we
 		 * just log a low-level debug message if it happens.
 		 */
+#if defined(__EMSCRIPTEN__) || defined(__wasi__)
+            HandleNotifyInterrupt();
+#else
 		if (SendProcSignal(pid, PROCSIG_NOTIFY_INTERRUPT, procnos[i]) < 0)
 			elog(DEBUG3, "could not signal backend with PID %d: %m", pid);
+#endif
 	}
 
 	pfree(pids);
