--- postgresql-snapshot-12-13-2000/src/backend/port/dynloader/linux.h	Mon May 29 03:00:17 2000
+++ postgresql-snapshot/src/backend/port/dynloader/linux.h	Sun Feb  4 23:30:59 2001
@@ -32,7 +32,8 @@
 #endif
 #else
 /* #define		pg_dlopen(f)	dlopen(f, 1) */
-#define pg_dlopen(f)	dlopen(f, 2)
+/* #define pg_dlopen(f)	dlopen(f, 2) */
+#define pg_dlopen(f) dlopen(f, (RTLD_NOW|RTLD_GLOBAL))
 #define pg_dlsym		dlsym
 #define pg_dlclose		dlclose
 #define pg_dlerror		dlerror
